allow variables in the color-theme definition file for workbench color

d-mahard opened the pull request:
<!-- Thank you for submitting a Pull Request. Please:
* Read our Pull Request guidelines:
  https://github.com/microsoft/vscode/wiki/How-to-Contribute#pull-requests
* Associate an issue with the Pull Request.
* Ensure that the code is up-to-date with the `main` branch.
* Include a description of the proposed changes and how to test them.
-->

## Description
For feature request described in #105247 

Mainly there are two new features proposed here

- Allow theme file to have a color palette, and the vars can be used to define a color for the **workbench** color item. _EDIT 2023-12-15_: also for syntax and semantic token.
- _EDIT 2023-12-15_ ~Add snippets and suggestion for the colors, which include all the defined variables in the palette~


This does **NOT** include the following

- _EDIT 2023-12-15_ ~Usage of the palette var for **token** colors (I am interested to explore this, but would like to hear feedback first)~
- Usage of the palette var in Setting [#56855](https://github.com/microsoft/vscode/issues/56855), _EDIT 2023-12-15_ ~as this needs to change the `colorRegistry`, which may affect many other things. (I would also love to explore this option, depending on the feedback)~

_EDIT 2023-12-15 (see comment for screenshots)_

## Some technical details

- The syntax using `$` is based on [this comment](https://github.com/microsoft/vscode/issues/105247#issuecomment-1246443785) by @aeschli 
- The parsing happens right when loading theme data. Any other service etc that uses color from the theme will always see the parsed hex, never the palette var.
- For the snippets/suggestion/validation, I cannot use `workbenchColorsSchemaId` because modifying it would affect many other things. So instead I create a new JSON Schema based on that ([here](https://github.com/d-mahard/vscode/blob/8a1ef6629883352dec9046cb428f3ba06cc1f5df/src/vs/workbench/services/themes/common/colorThemeSchema.ts#L234-L245)), so it will be isolated, only in theme file. However, if there is no palette, there will essentially be no modification to the schema, it is just copied as it is.
- _EDIT 2023-12-15_ ~For the snippets/suggestion/validation, the palette needs to be loaded first, meaning that if there is any change in the palette, VSCode needs to be restarted after the change. However, I think it is safe to assume that a palette change, especially the change of the _keys_, only happens rarely.~



## How to test

I prepared a unit test for the parsing in [themeColorParsing.test.ts](https://github.com/d-mahard/vscode/blob/8a1ef6629883352dec9046cb428f3ba06cc1f5df/src/vs/workbench/services/themes/test/node/themeColorParsing.test.ts)

aeschli commented:
I think the direction you are taking is good.
Please try to make the change as minimal as possible. That simplifies the review.

d-mahard commented:
Hi, thank you for the feedback! Sure, I'll try to minimize the changes.

d-mahard commented on src/vs/workbench/services/themes/common/colorThemeSchema.ts:
```diff
@@ -218,44 +219,101 @@ const textmateColorSchema: IJSONSchema = {
+	const themePaletteEnumSchema = schemaRegistry.getSchemaContributions().schemas[themePaletteEnumSchemaId];
+
+	if (themePaletteEnumSchema?.enum !== undefined && themePaletteEnumSchema.enum.length > 0) {
+		for (const key in workbenchColorSchemaForTheme.properties) {
```
This is the location where the workbench color schema is modified. 
If a palette is defined, then every color item (`workbenchColorSchemaForTheme.properties`) can either use the color-hex, or the var name from the palette.
If there is no palette, then the schema will exactly be the same with the original `workbenchColorsSchema`

d-mahard commented on src/vs/workbench/services/themes/common/colorThemeSchema.ts:
```diff
@@ -218,44 +219,101 @@ const textmateColorSchema: IJSONSchema = {
+				description: nls.localize('schema.semanticTokenColors', 'Colors for semantic tokens'),
+				$ref: tokenStylingSchemaId
+			},
+			colorPalette: {
```
Here the schema of the palette itself is defined.

d-mahard reviewed (commented):
@aeschli I tried to reduce the changes, but not so much could be removed.  Actually about half of the added lines are the unit test (two dummy theme files, and the unit test file itself). So for the logic, it is around 100 lines added and 30 removed.

To help with the review, I added some comments in this PR to explain the logic/design decision. 
Please see the comments below. 

Let me know if other information is needed.

aeschli commented on src/vs/workbench/services/themes/common/colorThemeData.ts:
```diff
@@ -709,16 +711,19 @@ async function _loadColorTheme(extensionResourceLoaderService: IExtensionResourc
 		result.semanticHighlighting = result.semanticHighlighting || contentValue.semanticHighlighting;
+		const colorPalette = contentValue.colorPalette;
+		const schemaRegistry = Registry.as<IJSONContributionRegistry>(JSONExtensions.JSONContribution);
+		schemaRegistry.registerSchema(themePaletteEnumSchemaId, { enum: Object.keys(colorPalette || {}) });
```
Every theme can have a different color palette. Reusing the same schema URI (themePaletteEnumSchemaId) for all themes is wrong.
- The last invocation of `_loadColorTheme` will define the content of the themePaletteEnumSchema. That doesn't have to be the theme the user is currently editing and for which a schema is needed. 
- Schemas should be available anytime and on demand. That should not depend on whether that theme has been loaded by VS Code.

IMO using JSON schemas to validate the use color variables doesn't work
If validation is needed, do this in code. We have extension `extensions/extension-editing`

aeschli commented:
Thanks @d-mahard for all the work!
I don't like the solution chosen to use a dynamic schema for the validation. Schemas are intended to be static resources, identified by an resource URI. You try to validate a document with a schema that is dynamically generated based on the document being validated. I don't see how his will work when opening a random theme file. Or when you open multiple theme files at the same time. 
If you want to add validation and completions, do that in code in the extension `extension-editing`. Or leave that to an external extension.
- remove the dynamic schema code. The only schema changes are to add the new property and allow `$/w+` besides color-hex
- the palette should also work in tokenColors and semantic colors.
- variables are local to a theme file. Make sure that the schemas for `workbench.colorCustomizations`, `editor.colorCustomizations` and `editor.semanticTokenColorCustomizations` are not affected and don't allow variables
- when using a color that is not defined (e.g. because of a typo in the color name), I suggest to default to `#f00` (red) as color, so that the problem is easy detectable

d-mahard commented:
thank you @aeschli for the review!
I understand your points. I will leave the dynamic schema validation/suggestion of the variable enumeration out of this PR. 
User will just need to satisfy the `$\w+` pattern beside the color-hex, and if it is not in the var list, it will default to red for easy debugging.

Regarding the following two points:

> - remove the dynamic schema code. The only schema changes are to add the new property and _allow `$\w+` besides color-hex_
> - Make sure that the schemas for `workbench.colorCustomizations`, `editor.colorCustomizations` and `editor.semanticTokenColorCustomizations` are not affected and don't allow variables

In my understanding, to allow both requirement we still need the dynamic schema. The reasons are the following:

Schemas for setting are controlled here. In line 84 below, we need to use `workbenchColorsSchemaId`, in which the color-hex format schema is defined.:
https://github.com/microsoft/vscode/blob/4fce419132bb1521d219d5c28e13c6098bda824e/src/vs/workbench/services/themes/common/themeConfiguration.ts#L81-L90



For the theme, the schema is (in the current `main` branch) as follows. In line 229 below, we need to also use `workbenchColorsSchemaId`, in which the color-hex format schema is defined.

https://github.com/microsoft/vscode/blob/4fce419132bb1521d219d5c28e13c6098bda824e/src/vs/workbench/services/themes/common/colorThemeSchema.ts#L222-L253



The `color-hex` limitation is defined in line 139 below, (and then the whole schema is registered in line 687 of the same file.)
https://github.com/microsoft/vscode/blob/4fce419132bb1521d219d5c28e13c6098bda824e/src/vs/platform/theme/common/colorRegistry.ts#L136-L153

We cannot simply just change the schema of the theme, because the `color-hex` limitation is not defined there. But if we add the `$\w+` to where it is defined (line 139 above), then it will definitely affect the setting.

So my proposal would still be using the dynamic schema, together with the scheduler, but removing the variable enum. So something like the following: 

```typescript
for (const key in workbenchColorSchemaForTheme.properties) {
		const property = workbenchColorSchemaForTheme.properties[key];
		delete property.format;

		property.anyOf = [
			{
				pattern: '^\\$\\w+', errorMessage: nls.localize('error.invalidFormat.colorEntryWithPalette', "Color must either in hex format (e.g. `#ffffff`) or one of the palette (e.g. `$accentName`)")
			}, // here, instead of using `$ref: themePaletteEnumSchemaId` like before, I just add the `$\w+` pattern
			{ format: 'color-hex' },
		];
	}
```

and of course `themePaletteEnumSchemaId` and all related logic everywhere will be removed.

(for completeness): The reason why I think even the scheduler is still needed is like I mentioned in my other comment:
> This scheduler is necessary because the color items are contributed not just from the "core" items in colorRegistry.ts but from some other locations, and it seems (from my observation at least) the loading/registering of those colors happen in parallel also with the loading of the theme itself. I tried calling the registerColorThemeSchemas() only once (without the scheduler), and as a result some of the colors in the theme file were linted as not allowed in the schema (i.e. the newly schema, modified from the workbench scheme). I concluded that when the function was called, not all color-items had already been registered. So it has to be repeated when there is new color registered.


Is my understanding correct or did I miss something here?

aeschli commented:
@d-mahard So as you say, we can no longer have a single schema to describe colors for `workbench.colorCustomizations` and for `colors` in the theme file.
I think the proper approach is to have two different schemas, e.g. 'vscode://schemas/workbench-settings-colors' and `'vscode://schemas/workbench-theme-colors'.

I would suggest to collect two different schemas in the ColorRegistry: https://github.com/microsoft/vscode/blob/c5874782c5127953ecb2daf5ba4e15819815e7a7/src/vs/platform/theme/common/colorRegistry.ts#L128

The trick with these schemas is that they are modified while new colors come in. So you can not just take a copy of it.

d-mahard commented:
@aeschli thank you for the hint!

I made the changes based on your suggestions:

- all error parsing now defaults to red
- dynamic schema validation/suggestion of the variable enumeration is no longer in this PR
- expanded the feature to cover tokenColors and semantic colors. The approach is the same with workbench color: parsing happens when theme data is loaded, so any other thing that uses the data will not see the variable.
- a new separate schema is collected from the colorRegistry for the workbench theme. 
- the same approach is also used for semantic token (separate schema for setting and theme), because I noticed that it is also updated every time there is a new semantic token rule added
- however for the token color, it seems not to be the case. So I just make a copy of the schema-for-setting and make the necessary changes for the theme schema
- I adjusted the unit test to cover also the parsing of token color and semantic token
- I wanted to add a unit test for the schema using something like `ajv`, but could not really figure out how module importing works in this project, so I decided to skip that one. If you can give me a hint I will try to create that.

In any case I tested the schema validation, see below:

![image](https://github.com/microsoft/vscode/assets/146386738/809e587e-779c-40d9-8de4-7b52258509eb)
_in settings, variable is not allowed_

![image](https://github.com/microsoft/vscode/assets/146386738/ac63d60b-5195-41a7-9e83-13a2dabbc0bd)
_in theme file, it is possible_

A little caveat:
The newly added warning (#200052) still works if we use hex, but when use variable there won't be any warning even when the color of that var is not transparent

aeschli commented:
@d-mahard Thanks for all your work! It's nicely done, I also tested it.
But now, seeing all the changes and talking about it with my co-workers, I'm hesitating to merge. I'm not happy with the added code complexity and added runtime memory around the schema generation. In the end, the color palettes are 'nice to have' but not essential. I'm not even sure that it will be good enough for ambitious theme authors, that probably want functions of colors. 
So again I think this should better done as a tool on top on top of theme files. A tool that generates color themes from it's own syntax.
I'm sorry that I didn't stop you earlier from doing all the work. I was hoping it's ending up to be a simple change, but I don't see any other approach than what you just implemented.
So I'm planing to close the PR and I hope for your understanding.

d-mahard commented:
@aeschli 
Thank you for taking the time to test and give feedback, really appreciated it.
I can totally understand the consideration. In any case, that was a good practice on my side. 👍
